core/src/org/javarosa/core/model/instance/TreeReference.java

```diff
@@ -279,57 +369,72 @@ public TreeReference parent(TreeReference parentRef) {
     }
 
 
-    //very similar to parent(), but assumes contextRef refers to a singular, existing node in the model
-    //this means we can do '/a/b/c + ../../d/e/f = /a/d/e/f', which we couldn't do in parent()
-    //return null if context ref is not absolute, or we parent up past the root node
-    //NOTE: this function still works even when contextRef contains INDEX_UNBOUND multiplicites... conditions depend on this behavior,
-    //  even though it's slightly icky
-    public TreeReference anchor(TreeReference contextRef) {
-        //TODO: Technically we should possibly be modifying context stuff here
-        //instead of in the xpath stuff;
+    /**
+     * Evaluate this reference in terms of a base absolute reference.
+     *
+     * For instance, anchoring ../../d/e/f to /a/b/c, results in  /a/d/e/f.
+     *
+     * NOTE: This function works when baseRef contains INDEX_UNBOUND
+     * multiplicites. Conditions depend on this behavior, but it is def
+     * slightly icky
+     *
+     * @param baseRef an absolute reference to be anchored to.
+     * @return null if base reference isn't absolute or there are too many
+     * '../'.
+     */
+    public TreeReference anchor(TreeReference baseRef) {
+        // TODO: Technically we should possibly be modifying context stuff here
+        // instead of in the xpath stuff;
 
         if (isAbsolute()) {
             return this.clone();
-        } else if (!contextRef.isAbsolute()) {
+        } else if (!baseRef.isAbsolute() ||
+                (refLevel > baseRef.size())) {
+            // non-absolute anchor ref or this reference has to many '../' for
+            // the anchor ref
             return null;
         } else {
-            TreeReference newRef = contextRef.clone();
-            int contextSize = contextRef.size();
-            if (refLevel > contextSize) {
-                return null; //tried to do '/..'
-            } else {
-                for (int i = 0; i < refLevel; i++) {
-                    newRef.removeLastLevel();
-                }
-                for (int i = 0; i < size(); i++) {
-                    newRef.add(data.elementAt(i).shallowCopy());
-                }
-                return newRef;
+            TreeReference newRef = baseRef.clone();
+            // remove a level from anchor ref for each '../'
+            for (int i = 0; i < refLevel; i++) {
+                newRef.removeLastLevel();
+            }
+            // copy level data from this ref to the anchor ref
+            for (int i = 0; i < size(); i++) {
+                newRef.add(this.data.elementAt(i).shallowCopy());
             }
+            return newRef;
         }
     }
 
-    //TODO: merge anchor() and parent()
-
```

Comment:
anchor and parent have pretty difference behavior when handling relative reference arguments, don't think they should eventually be merged